clients/lsp-csharp.el

```diff
@@ -448,6 +461,11 @@ using the `textDocument/references' request."
       (lsp-show-xrefs (lsp--locations-to-xref-items locations-found) nil t)
     (message "No references found")))
 
+(defun lsp-csharp--after-open-fn (&rest _)
```

Comment:
Ideally, this should run one time and before the server startup, not in after open. I propose you to set lsp-csharp-auto-update-server to nil after the update and also, ask the user to restart the server if they want new version to take effect. At some point we should move the update logic in the core.

Reply:
Yes, I agree with @yyoncho . I think this setting should be lsp-global, no need to instrument lsp-Csharp in particular IMHO.

Btw I have some changes that would move LSP-Sharp to use the standard LSP-mode package/dependency mechanism. @jcs090218 I would propose you rebase the changes after I make my PR, later this week, so a lot of custom version determination logic here would go away for good.

Reply:
And yes, hitting github APIs to check for new server version on every file open is too much.

can we hook into the “server started” event like @yyoncho has suggested?

Reply:
I suggest introducing an additional property in lsp--client named `can-update?`(or similar) which is a function. lsp-mode will call it only once per session and if the user confirms an update lsp-mode will call :download-server-fn with update=t.

Reply:
Okay, I have currently made this for check once per session by setting `lsp-csharp-auto-update-server` to `nil` after the first check.

> I suggest introducing an additional property in lsp--client named can-update?

That works for me. :)

> @jcs090218 I would propose you rebase the changes after I make my PR, later this week, so a lot of custom version determination logic here would go away for good.

@razzmatazz Sure, we can hang this PR until that point. I think you are referring to #2573?

Reply:
Ah, okay. I thought you were referring to the existing PR.

Reply:
The PR is here: #2748 – That PR, however, if merged would render this PR unmergable as there would be no more custom update code in lsp-csharp.

My humble opinion, @jcs090218, is that we should strive to make this "autoupdate" functionality part of lsp core so no custom code would be added to the clients, unless required for the extra functionality.

Not sure what @yyoncho feels about this conflict of PRs – which one should go.

Reply:
If this is going to core, I don't see point merging this PR. Close this now! :)